### PR TITLE
fix(TCOMP-2224): Support PREV_COLUMN_LIST / NEXT_COLUMN_LIST and BaseOnSchema in studio

### PR DESCRIPTION
**Pull request by wwang-talend**
### Requirements

* Any code change adding any logic **MUST** be tested through a unit test executed with the default build
* Any API addition **MUST** be done with a documentation update if relevant 

### Why this PR is needed?
https://jira.talendforge.org/browse/TCOMP-2224

### What does this PR adds (design/code thoughts)?
https://jira.talendforge.org/browse/TCOMP-2224

### CODE REVIEW

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import java.lang.annotation.Target;
+
+import org.talend.sdk.component.api.configuration.ui.meta.Ui;
+import org.talend.sdk.component.api.meta.Documentation;
+
+@Ui
+@Documentation("Mark a table column is filled by component schema auto, only for studio")
+@Retention(RUNTIME)
+@Target({ PARAMETER, FIELD })
+public @interface BaseOnSchema {
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java:32]: Thinking that Base**d**OnSchema naming is preferable.
```suggestion
public @interface BasedOnSchema {
```

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.talend.sdk.component.api.configuration.ui.meta.Ui;
+import org.talend.sdk.component.api.meta.Documentation;
+
+@Ui
+@Documentation("Mark a table column is filled by component schema auto, only for studio")
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java:29]:
```suggestion
@Documentation("Mark a table column filled by component's schema auto. Only for studio.")
```

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import java.lang.annotation.Target;
+
+import org.talend.sdk.component.api.configuration.ui.meta.Ui;
+import org.talend.sdk.component.api.meta.Documentation;
+
+@Ui
+@Documentation("Mark a ui option's suggestion values are filled by current component schema or pre connection line's schema auto, only for studio")
+@Retention(RUNTIME)
+@Target({ PARAMETER, FIELD })
+public @interface ColumnList {
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:32]: Isn't `ColumnReference`  a more accurate name?
```suggestion
public @interface ColumnReference {
```

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.talend.sdk.component.api.configuration.ui.meta.Ui;
+import org.talend.sdk.component.api.meta.Documentation;
+
+@Ui
+@Documentation("Mark a ui option's suggestion values are filled by current component schema or pre connection line's schema auto, only for studio")
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:29]:
```suggestion
@Documentation("Mark an UI suggestion value referencing a column from the current component schema or previous flow's schema. Only for studio.")
```

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+@Retention(RUNTIME)
+@Target({ PARAMETER, FIELD })
+public @interface ColumnList {
+
+    String value() default "curr";// or "pre"
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:34]: Shouldn't we prefer an enum? 
Also, prefer full naming. ie _current_, _previous_. Jira issue mention also next component's schema... (not sure it's relevant).

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java (hunk 3 above)
Comment (undx) [reply component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:32]: or `FieldReference`?
```suggestion
public @interface FieldReference {
```

**Review by undx — COMMENTED** [review]
some comments.

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java (hunk 3 above)
Comment (wwang-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:32]: this ui is a closedlist widget, which can choose one column from talend schema, so i think XXXList is better. And "Column" is more like a ui thing than Field.

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java (hunk 1 above)
Comment (wwang-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java:32]: done

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java (hunk 2 above)
Comment (wwang-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/BaseOnSchema.java:29]: done

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java (hunk 4 above)
Comment (wwang-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:29]: done

File: component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java (hunk 5 above)
Comment (wwang-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/configuration/ui/widget/ColumnList.java:34]: in javajet, also support pre schema one, so add that.

**Review by undx — DISMISSED** [review]
lgtm

**Review by pyzhou-talend — APPROVED** [review]
LGTM

**Pull request merged**